internal/credentials/credentials.go

```diff
@@ -0,0 +1,257 @@
+package credentials
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	"golang.org/x/crypto/bcrypt"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	lca_api "github.com/openshift-kni/lifecycle-agent/api/seedreconfig"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/image-based-install-operator/internal/certs"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultUser    = "kubeadmin"
+	kubeconfig     = "kubeconfig"
+	kubeadmincreds = "kubeadmincreds"
+)
+
+type Credentials struct {
+	client.Client
+	Log    logrus.FieldLogger
+	Scheme *runtime.Scheme
+}
+
+func (r *Credentials) EnsureKubeconfigSecret(ctx context.Context, cd *hivev1.ClusterDeployment, clusterInfo *lca_api.SeedReconfiguration) (*lca_api.KubeConfigCryptoRetention, error) {
+	existsAndValid, err := r.kubeconfigExistsAndValid(ctx, cd, clusterInfo)
+	if err != nil {
+		return nil, err
+	}
+	if existsAndValid {
+		return &clusterInfo.KubeconfigCryptoRetention, nil
+	}
+	// Generate cluster certs and create a new kubeconfig.
+	// TODO: handle user provided API and ingress certs
+	r.Log.Infof("Generating cluster crypto")
+	cm := certs.KubeConfigCertManager{}
+	if err := cm.GenerateAllCertificates(); err != nil {
```

Comment:
Do we ever want to call `GenerateAllCertificates` on an existing instance?
I feel like it would be safer for it to be `RegenerateAllCertificates` and throw out any existing ones at the start of the method.

Reply:
Sure, I don't think we need to change the name though.
GenerateAllCerts can just dump the existing ones

Reply:
Though when I think of it the existing certs is the only thing this object containes.
It dumping all the existing certs is essentially the same as creating a new object:
```
 func (r *KubeConfigCertManager) dumpExistingCertificates() {
	r = KubeConfigCertManager{}
}
```

Reply:
Sure, but it's just easier for a caller to create bugs if calling `GenerateAllCertificates` will append new certs.

I just don't see why that would ever be the desired behavior.